File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/molecules/button/roles/ButtonRole.kt
package com.perrystreet.woof.designsystem.atomic.molecules.button.roles

import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import com.perrystreet.woof.designsystem.atomic.atoms.icon.roles.IconColorRole
import com.perrystreet.woof.designsystem.atomic.atoms.text.roles.TextColorRole
import com.perrystreet.woof.designsystem.theme.Theme

enum class ButtonRole {
    Primary {
        @Composable
        override fun surfaceColor(): Color = Theme.colors.primary
        override val textColorRole = TextColorRole.OnPrimary
        override val loadingColorRole = IconColorRole.OnPrimary
    },
    Secondary {
        @Composable
        override fun surfaceColor(): Color = Theme.colors.surfaceContainerHigh
        override val textColorRole = TextColorRole.OnSurface
        override val loadingColorRole = IconColorRole.OnSurface
    },
    Destructive {
        @Composable
        override fun surfaceColor(): Color = Theme.colors.destructive
        override val textColorRole = TextColorRole.OnPrimary
        override val loadingColorRole = IconColorRole.OnPrimary
    },
    ;

    @Composable
    abstract fun surfaceColor(): Color
    abstract val textColorRole: TextColorRole
    abstract val loadingColorRole: IconColorRole
}

File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Molecules/Button/Roles/ButtonRole.swift
import SwiftUI

public enum ButtonRole {
    case primary
    case secondary
    case destructive

    public func surfaceColor(from theme: ThemeImplementing) -> Color {
        switch self {
        case .primary: theme.colors.primary
        case .secondary: theme.colors.surfaceContainerHigh
        case .destructive: theme.colors.destructive
        }
    }

    public var textColorRole: TextColorRole {
        switch self {
        case .primary: .onPrimary
        case .secondary: .onSurface
        case .destructive: .onPrimary
        }
    }

    public var loadingColorRole: IconColorRole {
        switch self {
        case .primary: .onPrimary
        case .secondary: .onSurface
        case .destructive: .onPrimary
        }
    }
}
